content/blog/2018-06-27-implementing-custom-sd.md

```diff
@@ -0,0 +1,190 @@
+---
+title: Implementing Custom Service Discovery
+created_at: 2018-06-27
+kind: article
+author_name: Callum Styan
+---
+
+## Implementing Custom Service Discovery
+
+Prometheus contains built in integrations for many service discovery (SD) systems such as Consul,
+Kubernetes, and public cloud providers such as Azure. However, we can’t provide integration
+implementations for every service discovery option out there. The Prometheus team is already stretched
+thin supporting the current set of SD integrations, so maintaining an integration for every possible SD
+option isn’t feasible. In many cases the current SD implementations have been contributed by people
+outside the team and then not maintained or tested well. We want to commit to only providing direct
+integration with service discovery mechanisms that we know we can maintain, and that work as intended.
+For this reason, there is currently a moratorium on new SD integrations.
+
+However, we know there is still a desire to be able to integrate with other SD mechanisms, such as 
+Docker Swarm. Recently a small code change plus an example was committed to the documentation 
+[directory](https://github.com/prometheus/prometheus/tree/master/documentation/examples/custom-sd)
+within the Prometheus repository for implementing a custom service discovery integration without having
+to merge it into the main Prometheus binary. The code change allows us to make use of the internal
+Discovery Manager code to write another executable that interacts with a new SD mechanism and outputs
+a file that is compatible with Prometheus' file\_sd. By co-locating Prometheus and our new executable
+we can configure Prometheus to read the file\_sd-compatible output of our executable, and therefore
+scrape targets from that service discovery mechanism. In the future this will enable us to move SD
+integrations out of the main Prometheus binary, as well as to move stable SD integrations that make
+use of the adapter into the Prometheus 
+[discovery](https://github.com/prometheus/prometheus/tree/master/discovery) package. 
+
+Popular integrations that are implemented with the adapter code may end up listed 
```

Comment:
Popularity isn't a requirement, I'd say that we list file_sd integrations such as those using this adapter on our integrations page.